yarn-project/archiver/src/archiver/l1_to_l2_message_store.ts

```diff
@@ -0,0 +1,79 @@
+import { Fr } from '@aztec/foundation/fields';
+import { L1ToL2Message } from '@aztec/types';
+
+/**
+ * A simple in-memory implementation of an L1 to L2 message store
+ * that handles message duplication.
+ */
+export class L1ToL2MessageStore {
+  /**
+   * A map containing the message key to the corresponding L1 to L2
+   * messages (and the number of times the message has been seen).
+   */
+  private store: Map<bigint, L1ToL2MessageAndCount> = new Map();
+
+  constructor() {}
+
+  addMessage(messageKey: Fr, msg: L1ToL2Message) {
+    const messageKeyBigInt = messageKey.value;
+    if (this.store.has(messageKeyBigInt)) {
+      this.store.get(messageKeyBigInt)!.count++;
+    } else {
+      this.store.set(messageKeyBigInt, { message: msg, count: 1 });
+    }
+  }
+
+  removeMessage(messageKey: Fr) {
+    const messageKeyBigInt = messageKey.value;
```

Comment:
It's a minor point but a couple of functions you are calling `has` before `get`. You can remove the `has` and just check if `get` returns undefined.

The only performance requirement on a Map is that lookups are better than O(N).